networkx/algorithms/shortest_paths/weighted.py

```diff
@@ -1044,26 +1076,28 @@ def johnson(G, weight='weight', new_weight=None):
     if not nx.is_weighted(G, weight=weight):
         raise nx.NetworkXError('Graph is not weighted.')
 
-    new_node = nx.utils.generate_unique_node()
-    G.add_weighted_edges_from((new_node, node, 0) for node in G.nodes())
+    dist = {}
+    pred = {}
+    for node in G:
+        dist[node] = 0
+        pred[node] = None
 
     # Calculate distance of shortest paths
-    dist = nx.bellman_ford(G, source=new_node, weight=weight)[1]
-
-    delete = False
-    if new_weight is None:
-        delete = True
-        new_weight = uuid.uuid1()
+    dist_bellman = _bellman_ford_relaxation(G, pred, dist, G.nodes(),
+                                            weight)[1]
 
-    for u, v, w in G.edges(data=True):
-        w[new_weight] = w[weight] + dist[u] - dist[v]
+    if G.is_multigraph():
+        get_weight = lambda u, v, data: min(eattr.get(weight, 1)
+                        for eattr in data.values()) + dist_bellman[u]\
+                                        - dist_bellman[v]
 
-    G.remove_node(new_node)
-    all_pairs_path = nx.all_pairs_dijkstra_path(G, weight=new_weight)
+    else:
+        get_weight = lambda u, v, data: data.get(weight, 1) + dist_bellman[u]\
+                                        - dist_bellman[v]
 
-    if delete:
-        for u, v, w in G.edges(data=True):
-            if new_weight in w:
-                w.pop(new_weight)
+    all_pairs_path = {}
+    for n in G:
+        paths = {n: [n]}  # dictionary of paths
+        all_pairs_path[n] = _dijkstra(G, n, get_weight, paths=paths)[1]
```

Comment:
Maybe simpler as a dictionary comprehension; then it's clear that it can be trivially parallelized (when the time comes to do that): `all_pairs = {v : _dijkstra(G, v, get_weight, paths={n: [n]})[1] for v in G}`.